### PR TITLE
add wasm extra proposal

### PR DESCRIPTION
**Pull request by zjg555543**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/wasm/keeper/keeper.go [hunk 1]
```diff
@@ -281,6 +281,55 @@ func (k Keeper) GetParams(ctx sdk.Context) types.Params {
...
+func (k *Keeper) UpdateGasRegister(ctx sdk.Context) {
+	gasFactor := k.GetGasFactor(ctx)
+	if gasFactor != k.gasRegister.GetGasMultiplier() {
+		k.gasRegister.UpdateGasMultiplier(gasFactor)
+	}
+	return
+}
+
+func (k *Keeper) modifyGasFactor(ctx sdk.Context, extra string) error {
+	result, err := types.NewActionModifyGasFactor(extra)
```
Comment (zjg555543) [inline x/wasm/keeper/keeper.go:307]: return sdk.Dec

File: x/wasm/types/keys.go [hunk 2]
```diff
@@ -114,6 +115,16 @@ func GetContractMethodBlockedListPrefix(contractAddr string) []byte {
 	return r
 }
 
+func GetGasFactorPrefix() []byte {
+	const gasFactor = "gasfactor"
```
Comment (zjg555543) [inline x/wasm/types/keys.go]: modify

File: x/gov/ante/ante.go [hunk 3]
```diff
@@ -50,6 +51,10 @@ func (ad AnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, ne
 				if err := ad.pk.CheckMsgSubmitProposal(ctx, msg); err != nil {
 					return ctx, err
 				}
+			case *wasmtypes.ExtraProposal:
```
Comment (BananaLF) [inline x/gov/ante/ante.go:54]: why is point to check, other is struct

File: x/wasm/keeper/keeper.go [hunk 4]
```diff
@@ -281,6 +281,56 @@ func (k Keeper) GetParams(ctx sdk.Context) types.Params {
 	return params
 }
 
+func (k Keeper) InvokeExtraProposal(ctx sdk.Context, action string, extra string) error {
+	switch action {
```
Comment (BananaLF) [inline x/wasm/keeper/keeper.go:285]: it 's wrong within ExtraProposal.ValidateBasic,  but it's correct here.

**Review by xzav3r — DISMISSED** [review]

**Review by xzav3r — APPROVED** [review]

**Review by KamiD — APPROVED** [review]

**Review by BananaLF — APPROVED** [review]

**Pull request merged**